design/object-bucket/object-bucket.md

```diff
@@ -0,0 +1,137 @@
+# ObjectBucketClaim CRD and Controller
+
+## Proposed Feature
+
+**A generalized method of S3 bucket provisioning through the implementation of an `ObjectBucketClaim` CRD and controller.** 
+
+Users will request buckets via the Kubernetes API and have returned ConfigMaps with connection information for that bucket.  
+
+One ConfigMap will be created per `ObjectBucketClaim` instance.  ConfigMap names will be deterministic and derived from the `ObjectBucketClaim` name.  This gives users the opportunity to define their Pod spec at the same time as the `ObjectBucketClaim`.  Additionally, the `ObjectBucketClaim` and Pod can be deployed at the same time thanks in-built synchronization in Kubernetes.
+
+Bucket deletion in the early stages of this controller design will not be addressed beyond cleaning up API objects.  Deletion of the bucket within the object store is left as an administrator responsibility.  This is to prevent the accidental destruction of data by users.
+
+## Design
+
+### Work Flow
+
+#### Assumptions
+
+- An object store has been deployed in a Kubernetes Cluster ([e.g. Rook-Ceph Object](https://rook.github.io/docs/rook/master/ceph-object.html))
+- A `Secret` (defined below) containing the Access Key and Secret Key has been created.
+- A `StorageClass` (defined below) has been created for the object store service.
+
+**Use Case: Provision a Bucket** 
+
+_As a Kubernetes user, I want leverage the Kubernetes API to create S3 buckets. I expect to get back the bucket connection information in a ConfigMap._
+ 
+![work flow](./ObjectBucketClaim.png)
+
+1. The Rook Operator detects a new `ObjectBucketClaim` instance.  
+    1. The operator uses `objectBucketClaim.spec.secretName` to get the S3 access keys secret.  
+    1. The operator uses `objectBucketClaim.spec.storageClassName` to get the `Service` endpoint of the object store.
+1. The operator uses the object store endpoint and access keys for an S3 "make bucket" call.
+1. The operator creates a ConfigMap in the namespace of the `ObjectBucketClaim` with relevant connection data of the bucket.
+1. An app Pod may then mount the Secret and the ConfigMap to begin accessing the bucket. 
+
+**Use Case:** Delete an Object Bucket
+
+_As a Kubernetes user, I want to delete `ObjectBucketClaim` instances and cleanup generated API resources._
+
+1. The user deletes the `ObjectBucketClaim` via `kubectl delete ...`.
+1. The `ObjectBucketClaim` is marked for deletion and left in the foreground.
```

Comment:
In this case, we won't be setting the finalizer ourselves.  Instead, we'll set and `ownerReference` with `blockOwnerDeletion: true` in the configMap to make it a child of the bucket claim.  The kubelet will be responsible for finalizers.  This keeps configMaps from being orphaned when the claim is deleted.

`blockOwnerDeletion` is more of a cherry on top since the configMap isn't blocked from deletion when referenced by a pod (as env var or volume).